src/Map.cpp

```diff
@@ -84,49 +88,90 @@ void Map::WriteToFile(std::string file_name) const {
-        sizeof (bounds.width)
-    );
+    for (const auto &tile_layer: tile_layers) {
+        int room_tile_count = tile_layer.tiles.size();
+        wf.write(reinterpret_cast<const char *>(&room_tile_count), sizeof (room_tile_count));
 
-    wf.write(
-        reinterpret_cast<const char *>(&bounds.height), 
-        sizeof (bounds.height)
-    );
+        for(const auto &t: tile_layer.tiles) {
+            wf.write(
+                reinterpret_cast<const char *>(&t.rotation), 
+                sizeof (t.rotation)
+            );
 
-    int room_tile_count = tiles->size();
-    wf.write(reinterpret_cast<const char *>(&room_tile_count), sizeof (room_tile_count));
+            wf.write(
+                reinterpret_cast<const char *>(&t.tile_map_index), 
+                sizeof (t.tile_map_index)
+            );
 
+            wf.write(
+                reinterpret_cast<const char *>(&t.x), 
+                sizeof (t.x)
+            );
 
-    for(const auto &t: *tiles) {
-        wf.write(
-            reinterpret_cast<const char *>(&t.rotation), 
-            sizeof (t.rotation)
-        );
+            wf.write(
+                reinterpret_cast<const char *>(&t.y), 
+                sizeof (t.y)
+            );
+        }
 
-        wf.write(
-            reinterpret_cast<const char *>(&t.tile_map_index), 
-            sizeof (t.tile_map_index)
-        );
+    }
 
-        wf.write(
-            reinterpret_cast<const char *>(&t.x), 
-            sizeof (t.x)
-        );
+    int entity_size = entities.size();
+    std::cout << entity_size << std::endl;
+    wf.write(reinterpret_cast<const char *>(&entity_size), sizeof (entity_size));
 
-        wf.write(
-            reinterpret_cast<const char *>(&t.y), 
-            sizeof (t.y)
-        );
+    std::for_each(entities.begin(), entities.end(), [&wf](Entity entity){
+        EntityType type = entity.GetEntityType();
+        //typedef std::underlying_type<EntityType>::type utype;
+        //int i = static_cast<utype>(type);
+
+        wf.write(reinterpret_cast<const char *>(&type), sizeof (type));
+
+        int x = entity.GetTransform().x;
+        wf.write(reinterpret_cast<const char *>(&x), sizeof (x));
+
+        int y = entity.GetTransform().y;
+        wf.write(reinterpret_cast<const char *>(&y), sizeof (y));
+    });
+
+}
+
+sf::IntRect Map::GetBounds() {
+    return bounds;
+}
+
+std::vector<TileLayer>& Map::GetTileLayers() {
+    return tile_layers;
+}
+
+void Map::AddTile(int index, MapTile tile) {
+    auto found = std::find_if(
+        tile_layers[index].tiles.begin(), 
+        tile_layers[index].tiles.end(), 
+        [tile](const auto &t) {
+            return t.x == tile.x && t.y == tile.y;
+        }
+    );
+
+    if (found != tile_layers[index].tiles.end()) {
+        tile_layers[index].tiles.erase(found);
     }
+    tile_layers[index].tiles.push_back(tile);
+}
+
+void Map::AddEntity(Entity entity) {
+    entities.push_back(entity);
```

Comment:
Should this check whether there is already an entity at this point (similar to `AddTile`)?

Reply:
Good question. I don't think we will need to do this as there are times I would imagine that we would have more than one entity on the same tile.